zeshel/retrieval/data.py

```diff
@@ -0,0 +1,469 @@
+            with open(os.path.join(doc_path, domain + '.json')) as f:
+                for line in f:
+                    field = json.loads(line)
+                    page_id = field['document_id']
+                    doc[page_id] = field
+        return doc
+
+    train_doc = load_entities(train_domain)
+    heldout_train_doc = load_entities(heldout_train_domain)
+    heldout_train_unseen_doc = load_entities(heldout_train_unseen_domain)
+    val_doc = load_entities(val_domain)
+    test_doc = load_entities(test_domain)
+
+    return samples_train, samples_heldout_train_seen, \
+           samples_heldout_train_unseen, samples_val, samples_test, \
+           train_doc, heldout_train_doc, heldout_train_unseen_doc, \
+           heldout_train_unseen_doc, val_doc, test_doc
+
+
+def get_all_entity_hiddens(en_loader, model, store_en_hiddens=False,
+                           en_hidden_path=None):
+    """
+    :param en_loader: entity data loader
+    :param model: current model
+    :param store_en_hiddens: store entity hiddens?
+    :param en_hidden_path: entity hidden states store path
+    :return: 
+    """
+    model.eval()
+    all_en_embeds = []
+    with torch.no_grad():
+        for i, batch in enumerate(en_loader):
+            if hasattr(model, 'module'):
+                en_embeds = (
+                    model.module.encode(None, None, None, None, batch[0],
+                                        batch[1])[
+                        2]).detach().cpu()
+            else:
+                en_embeds = (
+                    model.encode(None, None, None, None, batch[0],
+                                 batch[1])[
+                        2]).detach().cpu()
+            if store_en_hiddens:
+                file_path = os.path.join(en_hidden_path, 'en_hiddens_%s.pt' % i)
+                torch.save(en_embeds, file_path)
+            else:
+                all_en_embeds.append(en_embeds)
+    if store_en_hiddens:
+        return None
+    else:
+        all_en_embeds = torch.cat(all_en_embeds, dim=0)
+        return all_en_embeds
+
+
+def get_hard_negative(mention_loader, model,
+                      num_hards, len_en_loader,
+                      device, distribution_sampling=False,
+                      exclude_golds=True,
+                      too_large=False,
+                      all_candidates_embeds=None,
+                      en_hidden_path=None, adjust_logits=False,
+                      smoothing_value=1.0):
+    model.eval()
+    if not too_large:
+        if hasattr(model, 'module'):
+            model.module.evaluate_on = True
+            model.module.candidates_embeds = all_candidates_embeds
+        else:
+            model.evaluate_on = True
+            model.candidates_embeds = all_candidates_embeds
+    hard_indices = []
+    all_candidates_probs = []
+    with torch.no_grad():
+        for i, batch in tqdm(enumerate(mention_loader)):
+            if not too_large:  # entity embeddings are too large for memory
```

Comment:
Again just a minor comment (no need for change), but it's perhaps a better practice to pass an argument batch_size_small=-1 and change this to "if batch_size_small < 0:", so that we can process more than 1 entity.